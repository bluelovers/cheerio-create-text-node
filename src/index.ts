/**
 * Created by user on 2017/9/1/001.
 */
import type { AnyNode, Cheerio, AnyNode as ICheerioNode } from 'cheerio';
import type { Text as ICheerioTextNode, Text } from 'domhandler';
import { CheerioAPI } from 'cheerio';

export type { ICheerioNode, ICheerioTextNode }

export interface ICheerioTextObject
{
	type: 'text';
	data: string;
}

declare module 'cheerio'
{
	interface Cheerio<T>
	{
		createTextNode(text: string | AnyNode | unknown): Cheerio<Text>
	}

	interface CheerioAPI
	{
		createTextNode(text: string | AnyNode | unknown): Cheerio<Text>
	}
}

export function use<T extends Cheerio<any> | CheerioAPI>($: T): T & {
	createTextNode(): Text
}
{
	// @ts-ignore
	$.prototype && ($.prototype.createTextNode = createTextNodeCheerio);
	// @ts-ignore
	$.__proto__ && ($.__proto__.createTextNode = createTextNodeCheerio);

	return $ as any;
}

export function createTextNode(text: string | AnyNode | unknown): ICheerioTextObject
{
	return {
		type: 'text',
		// @ts-ignore
		data: (typeof text !== 'string' && typeof text.text == 'function') ? (text as Cheerio<any>).text() : text.toString(),
	};
}

export function createTextNodeCheerio(this: Cheerio<any>, text: string | AnyNode | unknown): Cheerio<Text>
{
	// @ts-ignore
	return ((this.prototype || this.__proto__) as any as Cheerio<any>)._make(createTextNode(text));
}

export default createTextNode;
