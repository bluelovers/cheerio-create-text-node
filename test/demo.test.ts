/**
 * Created by user on 2017/8/27/027.
 */

import { createTextNode, use } from '../src/index';

import cheerio, { Cheerio, CheerioAPI } from 'cheerio';
import { basename, extname } from 'path';

describe(basename(__filename, extname(__filename)), () =>
{
	let $: CheerioAPI;

	beforeEach(() =>
	{
		$ = cheerio.load('<ul>  <li></li>  </ul>');
		use($);
	});

	it(`test:append`, () =>
	{
		let len = $('ul').contents().length;

		let text = $.createTextNode(123);
		$('ul').append(text);

		expect($('ul').contents().length).toEqual(len + 1);
		expect($('ul').html()).toEqual('  <li></li>  123');
	});

	it(`test:appendTo`, () =>
	{
		let len = $('ul').contents().length;

		let text = $.createTextNode(123);
		text.appendTo($('ul'));

		expect($('ul').contents().length).toEqual(len + 1);
		expect($('ul').html()).toEqual('  <li></li>  123');
	});

	it(`test:append2`, () =>
	{
		let len = $('ul').contents().length;

		let text = createTextNode(123);
		$('ul').append(text as any);

		expect($('ul').contents().length).toEqual(len + 1);
		expect($('ul').html()).toEqual('  <li></li>  123');

		expect($('ul')).toMatchSnapshot();
	});

	it(`createTextNode return object`, () =>
	{
		let text = createTextNode(123);

		expect(text).toEqual({
			type: 'text',
			data: '123',
		});
	});

	it(`$.createTextNode instanceof cheerio`, () =>
	{
		let text = $.createTextNode(123);

		expect(text).toMatchSnapshot();
		expect(text[0]).toMatchObject({
			type: 'text',
			data: '123',
		});

		expect(text.cheerio).toMatchSnapshot();

		// @ts-ignore
		expect(text.nodeType).toMatchSnapshot();
	});

	it(`$('ul').createTextNode instanceof cheerio`, () =>
	{
		let text = $('ul').createTextNode(123);

		expect(text).toMatchSnapshot();
		expect(text[0]).toMatchObject({
			type: 'text',
			data: '123',
		});
	});

	it(`$.createTextNode(cheerio dom)`, () =>
	{
		let text = $.createTextNode(123);
		let text2 = $.createTextNode(text);

		expect(text.text()).toEqual(text2.text());
		expect(text).toMatchSnapshot();
		expect(text[0]).toMatchObject({
			type: 'text',
			data: '123',
		});
		expect(text2).toMatchSnapshot();
		expect(text2[0]).toMatchObject({
			type: 'text',
			data: '123',
		});
	});

});
