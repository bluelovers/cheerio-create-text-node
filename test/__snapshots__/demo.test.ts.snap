// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`demo.test $('ul').createTextNode instanceof cheerio 1`] = `
LoadedCheerio {
  "0": Object {
    "data": "123",
    "type": "text",
  },
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body>
            <ul>
                
              <li />
                
            </ul>
          </body>
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
  "prevObject": LoadedCheerio {
    "_make": [Function],
    "_parse": [Function],
    "_render": [Function],
    "constructor": [Function],
    "createTextNode": [Function],
  },
}
`;

exports[`demo.test $.createTextNode instanceof cheerio 1`] = `
LoadedCheerio {
  "0": Object {
    "data": "123",
    "type": "text",
  },
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body>
            <ul>
                
              <li />
                
            </ul>
          </body>
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
  "prevObject": LoadedCheerio {
    "_make": [Function],
    "_parse": [Function],
    "_render": [Function],
    "constructor": [Function],
    "createTextNode": [Function],
  },
}
`;

exports[`demo.test $.createTextNode instanceof cheerio 2`] = `"[cheerio object]"`;

exports[`demo.test $.createTextNode instanceof cheerio 3`] = `undefined`;

exports[`demo.test $.createTextNode(cheerio dom) 1`] = `
LoadedCheerio {
  "0": Object {
    "data": "123",
    "type": "text",
  },
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body>
            <ul>
                
              <li />
                
            </ul>
          </body>
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
  "prevObject": LoadedCheerio {
    "_make": [Function],
    "_parse": [Function],
    "_render": [Function],
    "constructor": [Function],
    "createTextNode": [Function],
  },
}
`;

exports[`demo.test $.createTextNode(cheerio dom) 2`] = `
LoadedCheerio {
  "0": Object {
    "data": "123",
    "type": "text",
  },
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body>
            <ul>
                
              <li />
                
            </ul>
          </body>
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
  "prevObject": LoadedCheerio {
    "_make": [Function],
    "_parse": [Function],
    "_render": [Function],
    "constructor": [Function],
    "createTextNode": [Function],
  },
}
`;

exports[`demo.test test:append2 1`] = `
LoadedCheerio {
  "0": <ul>
      
    <li />
      
    Object {
      "data": "123",
      "next": null,
      "parent": <ul>
          
        <li />
          
        [Circular]
      </ul>,
      "prev":   ,
      "type": "text",
    }
  </ul>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body>
            <ul>
                
              <li />
                
              Object {
                "data": "123",
                "next": null,
                "parent": <ul>
                    
                  <li />
                    
                  [Circular]
                </ul>,
                "prev":   ,
                "type": "text",
              }
            </ul>
          </body>
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
  "prevObject": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body>
            <ul>
                
              <li />
                
              Object {
                "data": "123",
                "next": null,
                "parent": <ul>
                    
                  <li />
                    
                  [Circular]
                </ul>,
                "prev":   ,
                "type": "text",
              }
            </ul>
          </body>
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
}
`;
